File: 3nd series/question1_(پیست اسکی)/S1.java
import java.util.LinkedList;
import java.util.Scanner;


class V
{
    int i, j;

    public V(int i, int j)
    {
        this.i = i;
        this.j = j;
    }


    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;

        if (!(obj instanceof V))
            return false;

        V hold = (V) obj;

        return (this.i == hold.i && this.j == hold.j);
    }
}


public class S1
{
    private static int n, m;
    private static Scanner in = new Scanner(System.in);



    public static boolean isInRange(int y, int x)
    {
        if (x < 0 || y < 0)
            return false;

        if (x >= m)
            return false;

        if (y >= n)
            return false;

        return true;
    }




    public static void main(String[] args) 
    {
        String line = in.nextLine();
        n = Integer.parseInt(line.split(" ")[0]);
        m = Integer.parseInt(line.split(" ")[1]);

        boolean[][] visited = new boolean[n][m];
        int[][] dis = new int[n][m];


        LinkedList<V> q = new LinkedList<>();
        
        
        for (int i = 0; i < n; i++)
        {
            line = in.nextLine();

            for (int j = 0; j < line.length(); j++)
            {
                visited[i][j] = false;

                if (line.charAt(j) == 'M')
                {
                    q.add(new V(i, j));
                    visited[i][j] = true;
                    dis[i][j] = 0;
                }

                else if (line.charAt(j) == 'X')
                {
                    visited[i][j] = true;
                    dis[i][j] = -1;
                }

                else
                    dis[i][j] = -1;
            }
        }


        while (q.size() != 0)
        {
            V s = q.removeFirst();

            if (isInRange(s.i-1, s.j) && !visited[s.i-1][s.j])
            {
                q.add(new V(s.i-1, s.j));
                dis[s.i-1][s.j] = dis[s.i][s.j] + 1;
                visited[s.i-1][s.j] = true;
            }

            if (isInRange(s.i+1, s.j) && !visited[s.i+1][s.j])
            {
                q.add(new V(s.i+1, s.j));
                dis[s.i+1][s.j] = dis[s.i][s.j] + 1;
                visited[s.i+1][s.j] = true;
            }

            if (isInRange(s.i, s.j-1) && !visited[s.i][s.j-1])
            {
                q.add(new V(s.i, s.j-1));
                dis[s.i][s.j-1] = dis[s.i][s.j] + 1;
                visited[s.i][s.j-1] = true;
            }
            if (isInRange(s.i, s.j+1) && !visited[s.i][s.j+1])
            {
                q.add(new V(s.i, s.j+1));
                dis[s.i][s.j+1] = dis[s.i][s.j] + 1;
                visited[s.i][s.j+1] = true;
            }           
        }


        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
                System.out.print(dis[i][j] + " ");

            System.out.println();
        }
    }
}

File: 3nd series/question1_(پیست اسکی)/s1.py

def is_in_range(y, x):
    global m, n

    if x < 0 or y < 0:
        return False
    if x >= m:
        return False
    if y >= n:
        return False

    return True





if __name__ == "__main__":
    
    line = input().split()
    n, m = int(line[0]), int(line[1])
    visited = []
    dis = []

    q = []

    
    for i in range(n):
        line = input()
        visited.append([])
        dis.append([])

        for j in range(len(line)):

            visited[i].append(False)
            dis[i].append(-1)
            
            if line[j] == 'M':
                q.append((i, j))
                visited[i][j] = True
                dis[i][j] = 0

            elif line[j] == 'X':
                visited[i][j] = True
                dis[i][j] = -1
            

    while q:
        s = q.pop(0)

        if is_in_range(s[0]-1, s[1]) and (not visited[s[0]-1][s[1]]):
            q.append((s[0]-1, s[1]))
            dis[s[0]-1][s[1]] = dis[s[0]][s[1]] + 1
            visited[s[0]-1][s[1]] = True

        if is_in_range(s[0]+1, s[1]) and (not visited[s[0]+1][s[1]]):
            q.append((s[0]+1, s[1]))
            dis[s[0]+1][s[1]] = dis[s[0]][s[1]] + 1
            visited[s[0]+1][s[1]] = True

        if is_in_range(s[0], s[1]-1) and (not visited[s[0]][s[1]-1]):
            q.append((s[0], s[1]-1))
            dis[s[0]][s[1]-1] = dis[s[0]][s[1]] + 1
            visited[s[0]][s[1]-1] = True

        if is_in_range(s[0], s[1]+1) and (not visited[s[0]][s[1]+1]):
            q.append((s[0], s[1]+1))
            dis[s[0]][s[1]+1] = dis[s[0]][s[1]] + 1
            visited[s[0]][s[1]+1] = True
        

    for i in range(n):
        print(*dis[i])




    
    


            


    